src/Kestrel.Core/Internal/HttpConnectionMiddleware.cs

```diff
@@ -0,0 +1,111 @@
+    {
+        private static long _lastFrameConnectionId = long.MinValue;
+
+        private readonly IList<IConnectionAdapter> _connectionAdapters;
+        private readonly ServiceContext _serviceContext;
+        private readonly IHttpApplication<TContext> _application;
+
+        public HttpConnectionMiddleware(IList<IConnectionAdapter> adapters, ServiceContext serviceContext, IHttpApplication<TContext> application)
+        {
+            _serviceContext = serviceContext;
+            _application = application;
+
+            // Keeping these around for now so progress can be made without updating tests
+            _connectionAdapters = adapters;
+        }
+
+        public Task OnConnectionAsync(ConnectionContext connectionContext)
+        {
+            // We need the transport feature so that we can cancel the output reader that the transport is using
+            // This is a bit of a hack but it preserves the existing semantics
+            var transportFeature = connectionContext.Features.Get<IConnectionTransportFeature>();
+
+            var frameConnectionId = Interlocked.Increment(ref _lastFrameConnectionId);
+
+            var frameConnectionContext = new FrameConnectionContext
+            {
+                ConnectionId = connectionContext.ConnectionId,
+                FrameConnectionId = frameConnectionId,
+                ServiceContext = _serviceContext,
+                PipeFactory = connectionContext.PipeFactory,
+                ConnectionAdapters = _connectionAdapters,
+                Transport = connectionContext.Transport,
+                Application = transportFeature.Application
+            };
+
+            var connectionFeature = connectionContext.Features.Get<IHttpConnectionFeature>();
+
+            if (connectionFeature != null)
+            {
+                if (connectionFeature.LocalIpAddress != null)
+                {
+                    frameConnectionContext.LocalEndPoint = new IPEndPoint(connectionFeature.LocalIpAddress, connectionFeature.LocalPort);
+                }
+
+                if (connectionFeature.RemoteIpAddress != null)
+                {
+                    frameConnectionContext.RemoteEndPoint = new IPEndPoint(connectionFeature.RemoteIpAddress, connectionFeature.RemotePort);
+                }
+            }
+
+            var connection = new FrameConnection(frameConnectionContext);
+
+            // The order here is important, start request processing so that
+            // the frame is created before this yields. Events need to be wired up
+            // afterwards
+            var processingTask = connection.StartRequestProcessing(_application);
+
+            // Wire up the events an forward calls to the frame connection
+            // It's important that these execute synchronously because graceful
+            // connection close is order sensative (for now)
+            connectionContext.ConnectionAborted.ContinueWith((task, state) =>
```

Comment:
ContinueWith! I think I remember you claiming there's never a good excuse for this 😉